### ISSUE TITLE
Wildcard domains

### ISSUE THREAD

**Issue by Sexual** [caddyserver/ingress#55]
Is it possible to set up an ingress to support all domains?

E.g: `host: *` rather than `host: foo.com`

For my use case, I want to support automatic cert issuing for all domains, but the amount of domains is constantly changing and dynamic and can't be manually set in the standard K8s ingress host values.

**Comment by mholt** [comment]
(I can't answer for the maintainers, but I'll just note, in case it is helpful, that `*` is not a valid wildcard representation of `foo.com` -- you'd need either `*.com` or `*.*`, which, as certificate subjects, most browsers/clients reject.)

**Comment by Embraser01** [comment]
Hi 

I think what you search for is On Demand TLS together with a default backend. It is not implemented yet but it's the next thing I want to add 🙂

**Comment by Sexual** [comment]
@Embraser01 On-Demand TLS with a default backend sounds like just what I need!

Is there any ETA for this or any advice on where to get started to help implement this feature?

**Comment by Embraser01** [comment]
I'm looking to implement it before the end of the month, will update here when I've made some progress

**Comment by Sexual** [comment]
> I'm looking to implement it before the end of the month, will update here when I've made some progress

Any update regarding this? Thanks

**Comment by Embraser01** [comment]
> Any update regarding this? Thanks

Yes! I decided to work on an improved version of the controller and refactored a bunch of things. It also add support for OnDemand TLS. 

It's not merged yet as there is still some things to finish but the controller should be working.
The code is [here](https://github.com/caddyserver/ingress/pull/60) and a docker image of the latest commit (this morning) is available [here](https://hub.docker.com/r/caddy/ingress/tags?page=1&ordering=last_updated&name=pr-60) or [here](https://hub.docker.com/repository/docker/embraser01/caddy-ingress-controller/tags?page=1&ordering=last_updated).

No documentation yet but it's as easy as adding a few fields in the configmap ([JSON schema](https://github.com/caddyserver/ingress/blob/embraser01/rewrite-try/charts/caddy-ingress-controller/values.schema.json#L93-L150)).

**Comment by Sexual** [comment]
> > Any update regarding this? Thanks
> 
> Yes! I decided to work on an improved version of the controller and refactored a bunch of things. It also add support for OnDemand TLS.
> 
> It's not merged yet as there is still some things to finish but the controller should be working.
> The code is [here](https://github.com/caddyserver/ingress/pull/60) and a docker image of the latest commit (this morning) is available [here](https://hub.docker.com/r/caddy/ingress/tags?page=1&ordering=last_updated&name=pr-60) or [here](https://hub.docker.com/repository/docker/embraser01/caddy-ingress-controller/tags?page=1&ordering=last_updated).
> 
> No documentation yet but it's as easy as adding a few fields in the configmap ([JSON schema](https://github.com/caddyserver/ingress/blob/embraser01/rewrite-try/charts/caddy-ingress-controller/values.schema.json#L93-L150)).

Awesome work! Is this ready to be used in a live environment and are there installation instructions for the improved and refactored ingress?

**Comment by Embraser01** [comment]
> Awesome work! Is this ready to be used in a live environment and are there installation instructions for the improved and refactored ingress?

The project being still very young, I can't make promises. I can tell you that I've been using it in a live environment for some weeks now and it runs very nicely!
I didn't have time to update the documentation yet, but you can check out the chart folder and use helm to generate your manifests https://github.com/caddyserver/ingress/tree/66c52c682f497022f43ffb529def89c3a8ff3472/charts/caddy-ingress-controller

**Comment by Sexual** [comment]
> > Awesome work! Is this ready to be used in a live environment and are there installation instructions for the improved and refactored ingress?
> 
> The project being still very young, I can't make promises. I can tell you that I've been using it in a live environment for some weeks now and it runs very nicely!
> I didn't have time to update the documentation yet, but you can check out the chart folder and use helm to generate your manifests https://github.com/caddyserver/ingress/tree/66c52c682f497022f43ffb529def89c3a8ff3472/charts/caddy-ingress-controller

Do you have any insight into how to configure the default backend? I've tried deploying it and checking the logs, it's constantly checking an existing ingress' hosts and trying to issue certificates for them.

**Comment by Sexual** [comment]
@Embraser01  So after a lot of experimenting, I've gotten a lot further on this.

1. Having a wildcard for ALL hosts supporting automatic HTTPS isn't possible as the hostnames must be explicitly set: https://caddyserver.com/docs/automatic-https#activation 

> Any of the following will prevent automatic HTTPS from being activated, either in whole or in part: Not providing any hostnames or IP addresses in the config

2. The `proxyProtocol` is not working preventing actual production use. When trying to access `https://example.com` (a domain configured in the ingress and pointed to the external load balancer IP). It logs the following errors when trying to access via HTTPS:
```
[caddy-caddy-ingress-controller-7b5c7c9577-z7r8b] {"level":"debug","ts":1608994766.4413562,"logger":"http.stdlib","msg":"http: TLS handshake error from 10.128.0.61:8846: invalid signature"} 
[caddy-caddy-ingress-controller-7b5c7c9577-z7r8b] {"level":"debug","ts":1608994766.6447232,"logger":"http.stdlib","msg":"http: TLS handshake error from 10.128.0.61:22483: invalid signature"} 
[caddy-caddy-ingress-controller-7b5c7c9577-vt5qw] {"level":"debug","ts":1608994766.6778147,"logger":"http.stdlib","msg":"http: TLS handshake error from 10.36.5.1:16605: invalid signature"} 
[caddy-caddy-ingress-controller-7b5c7c9577-z7r8b] {"level":"debug","ts":1608994766.8814533,"logger":"http.stdlib","msg":"http: TLS handshake error from 10.36.4.1:59504: invalid signature"} 
[caddy-caddy-ingress-controller-7b5c7c9577-z7r8b] {"level":"debug","ts":1608994767.8734257,"logger":"http.stdlib","msg":"http: TLS handshake error from 10.36.4.1:31946: invalid signature"} 
[caddy-caddy-ingress-controller-7b5c7c9577-z7r8b] {"level":"debug","ts":1608994768.0778928,"logger":"http.stdlib","msg":"http: TLS handshake error from 10.128.0.61:11642: invalid signature"} 
[caddy-caddy-ingress-controller-7b5c7c9577-z7r8b] {"level":"debug","ts":1608994768.1217213,"logger":"http.stdlib","msg":"http: TLS handshake error from 10.128.0.61:5770: invalid signature"} 
[caddy-caddy-ingress-controller-7b5c7c9577-vt5qw] {"level":"debug","ts":1608994768.3253047,"logger":"http.stdlib","msg":"http: TLS handshake error from 10.36.5.1:57416: invalid signature"} 
```

When trying to access via HTTP, it provides a white error page with the text: `400 Bad Request`

**Comment by Embraser01** [comment]
1. Having a wildcard for ALL hosts supporting automatic HTTPS isn't possible as the hostnames must be explicitly set: [caddyserver.com/docs/automatic-https#activation](https://caddyserver.com/docs/automatic-https#activation)

Could you provide an ingress `.yml` and the configmap `.yml` files that would help me reproduce your issue? 
And if you also can provide us the applied `config.json` (it's logged by the controller).

2. The `proxyProtocol` is not working preventing actual production use. When trying to access `https://example.com` (a domain configured in the ingress and pointed to the external load balancer IP). It logs the following errors when trying to access via HTTPS:

For now, when PROXY protocol is enabled, it prevent any connection that do not use PROXY protocol. I don't know exactly your setup but I know that in order to enable PROXY Protocol in AWS, I had to make sure the load balancer in front of Caddy is in `ip-mode`.

With ip mode enabled: LoadBalancer -> Caddy Ingress Controller Pod
Without it: LoadBalancer -> Node -> Caddy Ingress Controller Pod

**Comment by kyranb** [comment]
Just confirming, thanks to #65 will on demand TLS for any host name be possible? As is currently possible when using Caddy standalone (not as an ingress).

**Comment by Sexual** [comment]
@Embraser01 It appears that this is due to the PROXY protocol not being something that is explicitly able to be set on GKE. https://projectcontour.io/guides/proxy-proto/

I'm not too sure how this works with ingress-nginx though as I have no issues with that.

Is there any workarounds to this issue?

**Comment by RickFoland** [comment]
I'd also like to confirm that OnDemand is supposed to be working. I spun up the ingress controller with `--set ingressController.config.onDemandTLS=true` and although it issues certs as expected, it serves a blank page instead of following the backend specified in my ingress resource.

**Comment by kosmoz** [comment]
I got it working. You have to create an ingress rule without a host, rather than a default backend:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: dummy
spec:
  ingressClassName: caddy
  rules:
    - http:
        paths:
          - pathType: Prefix
            path: "/"
            backend:
              service:
                name: dummy
                port:
                  number: 80
```

Also, it does not work unless you also configure `ingressController.config.onDemandAsk`.

I think this should be added to the documentation.
